docs/content/en/docs/Design/simd.md

```diff
@@ -0,0 +1,112 @@
+In addition, the passes make use of the `RotationAnalysis` and `TargetSlotAnalysis` analyses.
+
+### High-Level Flow
+* **Loop Unrolling** (`-full-loop-unroll`):
+    The implementation currently begins by unrolling all loops in the program to simplify the later passes.
+    See [#589](https://github.com/google/heir/issues/589) for a discussion on how this could be avoided.
+* **Canonicalization** (`-apply-folders -canonicalize`):
+    As the rotation-specific passes are very strict about the structure of the IR they operate on, we must first simplify away things such as tensors of constant values.
+    For performance reasons (c.f. comments in the `heirSIMDVectorizerPipelineBuilder` function in `heir-opt.cpp`), this must be done by first applying [folds](https://mlir.llvm.org/docs/Canonicalization/#canonicalizing-with-the-fold-method) before applying the full [canonicalization](https://mlir.llvm.org/docs/Canonicalization/).
+* **Main SIMD Rewrite** (`-insert-rotate -cse -canonicalize -cse`):
+    This pass rewrites arithmetic operations over `tensor.extract`-ed operands into SIMD operations over the entire tensor, rotating the (full-tensor) operands so that the correct elements interact.
+    For example, it will rewrite the following snippet (which computes `t2[4] = t0[3] + t1[5]`)
+    ```llvm
+    %0 = tensor.extract %t0[%c3] : tensor<32xi16>
+    %1 = tensor.extract %t1[%c5] : tensor<32xi16>
+    %2 = arith.addi %0, %1 : i16
+    %3 = tensor.insert %2 into %t2[%c4] : tensor<32xi16>
+    ```
+    to
+    ```llvm
+    %0 = tensor_ext.rotate %t0, %c31 : tensor<32xi16>, index
+    %1 = tensor_ext.rotate %t1, %c1 : tensor<32xi16>, index
+    %2 = arith.addi %0, %1 : tensor<32xi16>
+    ```
+    i.e., rotating `t0` down by one (31 = -1 (mod 32)) and  `t1` up by one to bring the elements at index 3 and 5, respectively, to the "target" index 4.
+    The pass uses the `TargetSlotAnalysis` to identify the appropriate target index (or ciphertext "slot" in FHE-speak).
+    See [Insert Rotate Pass](#insert-rotate-pass) below for more details. This pass is roughly equivalent to the `-batching` pass in the original HECO implementation.
+
+    Doing this rewrite by itself does not represent an optimization, but if we consider what happens to the corresponding code for other indices (e.g.,  `t2[5] = t0[4] + t1[6]`), we see that the pass transforms expressions with the same relative index offsets into the exact same set of rotations/SIMD operations, so the following [Common Subexpression Eliminiation (CSE)](https://en.wikipedia.org/wiki/Common_subexpression_elimination) will remove redundant computations.
+    We apply CSE twice, once directly (which creates new opportunities for canonicalization and folding) and then again after that canonicalization.
```

Comment:
Is it worth explaining what those canoncalization patterns are?

Reply:
I added a brief "tensor text canonicalizations" section to the document and referenced it from here, but I feel like this information really out to go to https://heir.dev/docs/dialects/tensorext/, I'm just not sure where I'd have to put it for it to appear there.

Reply:
I agree it feels like it should belong to canonicalization in that dialect. I bet we can just add a README.md to each Dialect subfolder and then append that to the markdown generated https://github.com/google/heir/blob/15e510cf97874bac4f7901bc0fe5556e01bc7294/.github/workflows/copy_tblgen_files.sh#L64

right below the header: https://heir.dev/docs/dialects/tensorext/

Maybe the secret design section should actually go in like that too, and maybe each dialect should optionally have  a Design/MoreInfo subpage. We can discuss on an issue